pkg/gadgettracermanager/pubsub/pubsub.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pubsub
+
+import (
+	"sync"
+	"time"
+
+	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type PubSubEvent struct {
+	Deleted   bool
+	Container pb.ContainerDefinition
+	Wg        *sync.WaitGroup
+}
+
+// GadgetPubSub provides a synchronous publish subscribe mechanism for gadgets
+// to be informed of container creation and deletion. It needs to be
+// synchronous so that gadgets have time to attach their tracer before the
+// container is started.
+type GadgetPubSub struct {
+	mu sync.RWMutex
+
+	// subs contains a list of subscribers
+	subs []chan PubSubEvent
+}
+
+func NewGadgetPubSub() *GadgetPubSub {
+	return &GadgetPubSub{
+		subs: make([]chan PubSubEvent, 0),
+	}
+}
+
+func (g *GadgetPubSub) Subscribe() chan PubSubEvent {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	ch := make(chan PubSubEvent)
+	g.subs = append(g.subs, ch)
+	return ch
+}
+
+func (g *GadgetPubSub) Unsubscribe(ch chan PubSubEvent) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	keep := make([]chan PubSubEvent, 0)
+	for _, i := range g.subs {
+		if ch != i {
+			keep = append(keep, i)
+		}
+	}
+	g.subs = keep
+}
+
+func (g *GadgetPubSub) Publish(deleted bool, container pb.ContainerDefinition) {
+	g.mu.RLock()
+	defer g.mu.RUnlock()
+
+	var wg sync.WaitGroup
+
+	for _, ch := range g.subs {
+		wg.Add(1)
+		go func(ch chan PubSubEvent) {
+			ch <- PubSubEvent{
+				Deleted:   deleted,
+				Container: container,
+				Wg:        &wg,
+			}
+		}(ch)
```

Comment:
What about if `Subscribe()` receives a callback instead of returning a channel?, by doing so the caller won't have to worry about calling `event.Wg.Done()`. Do you see advantages on using a channel instead?

Reply:
By using a channel, we ensure that the publisher (gadget tracer manager) and the subscriber (a gadget) have their own go routine. If a gadget has performance problems (compiling a BPF program taking a while..), I can timeout. But it is arguable whether it is a good thing... Do you think I should try with a callback?

Reply:
I think we can do the same but controlling everything from the publisher point of view. 

Something like: 
```go
	for _, sub := range g.subs {
		wg.Add(1)
		go func() {
			event := PubSubEvent{
				Deleted:   deleted,
				Container: container,
			}
			sub.Callback(event)
			wg.Done()
		}()
	}
```

Reply:
Functions are not comparable so they cannot be used as a key in the `subs` map. I will try to use a `interface{}` for the key.